BlackCubes/react-challenge-project-sep-trial
# Login and Signup Systems

## Changes
1. Created a global error handler on the Server for better error debugging and messages.
2. Created validations on the Server to handle the authentication/user inputs.
3. Created a MongoDB model with Mongoose.js on the Server for the user with password encryption.
4. Created routes and middlewares on the Server for authentication/user.
5. Created an error route on the Server for any unknown routes.
6. Created and expanded the actions, types, and reducers for login and signup authentications on the Client.
7. Created custom error and validation handlers on the Client for the forms.
6. Refactored the Login components on making them functional components while also implementing redux on the Client.
7. Created a link on the Login component to send the user to the `/signup` path for registering.
7. Created Signup components and implemented it with redux and react-router-dom on the Client.

## Purpose
There should be a user account in place on the Server for both logging in and signing up, and the Client should decide whether the user has submitted their inputs successfully as well if they should proceed to a new path.

Additionally, there should be a global error handler on the Server that displays friendly and useful error messages for the developer to debug, and for the user to submit proper inputs.

## Approach
In order for there to be a user account, a model with the implementation of Mongoose.js and MongoDB was created that includes proper password encryption by the implementation of an NPM package called `bcryptjs`. Routes and middlewares were created for the user-authentication, and along with this, handler validations were also created to deal with the user logging in and signing up with useful error messages.

To better handle any errors from the Server, and so it would not crash it, a global error handler was created which provides debugging messages for the developer, and useful messages for the user on what type of inputs they should send. Also, when the Server is being run during LIVE production, shorter error messages would be displayed so that nothing important gets leaked and likewise, when the Server is being run during development, more descriptive error messages would be shown.

Another implementation on the Server was handling any unknown routes besides the ones created. This would display a friendly error message in JSON format without crashing the server utilizing the global error handler.

For the Client, new actions, types, and reducers were created to handle the signup, error, and success on the user-authentication which all communicates to the newly expanded Server. Custom validations and error checking were also created for the forms to keep code clean. Signup and its form was created with the implementation of redux on handling user signup, and the login and its form was restructured to better use redux, react-router-dom, and functional components.

In all, this brings a very exciting and dynamic feature to the website and for the user.

## Testing
1. Pull changes.
2. On the parent directory, run `docker compose down && docker compose up --build` on your terminal.
3. Client:
  - Go to the `/login` path on your browser and make sure you are not being redirected after a failed login, and make sure the error messages appear underneath the inputs.
  - Click on the "Sign up now" link below the "Login" button and make sure you are redirected to the `/signup` path.
  - Create a account while making sure any errors appears underneath the inputs.
  - Make sure you are not redirected with incorrect inputs, and likewise, make sure you are redirected to the `/login` path upon successful inputs.
  - On the `/login` path after successfully completing the registration, login in and make sure you are redirected to the `/view-orders` path upon.
  - Use the back button on your browser and make sure you are not back on the `/login` path.
4. Server:
  - Create endpoints of `/api/login` and `/api/signup` on Postman, and both should be `POST` methods.
  - Underneath the request URL, select the "Body" option and then select `raw`.
  - On the right side next to the "GraphQL" option, click on the blue "Text" option and choose "JSON".
  - Inside the text-box underneath the options, create an empty object `{ }`.
  - Expand the empty object and add properties of `"email"` and `"password"` for the `/api/login` route; or add `"email"`, `"password"`, and `"password_confirmation"` for the `/api/signup` route.
  - Intentionally add bad values (or none) to the two routes and make sure you get a detailed JSON error response since the server is running on development.
  - Follow what the JSON error responses tells you on what to type for the correct values, and once you add correct values, make sure you get a success JSON response.

## Screenshot
#### Signup Link
![signup-link](https://user-images.githubusercontent.com/29642735/133870797-d42eff61-399b-4a51-8901-cd455f4ee541.gif)

#### Signup Fail
![signup-fail](https://user-images.githubusercontent.com/29642735/133870807-9838065e-5a74-4c76-8c02-38dcf7b1c021.gif)

#### Signup Success
![signup-success](https://user-images.githubusercontent.com/29642735/133870816-2bc51898-76d8-4cbc-82bb-737a1d72fa8d.gif)

#### Login Fail
![login-fail](https://user-images.githubusercontent.com/29642735/133870825-cab8965e-636b-4fd1-8fb7-0d2515897e79.gif)

#### Login Success
![login-success](https://user-images.githubusercontent.com/29642735/133870835-780589d8-8697-43cd-9752-f27fed7f9a32.gif)

#### Back Button After Logging In
![back-button](https://user-images.githubusercontent.com/29642735/133870850-01d1c893-f2e0-49c9-8670-442683f92c35.gif)


Closes #11 
Closes #19
